### ISSUE TITLE
Serialized Img?

### ISSUE THREAD

**Issue by larrylindsey** [imglib/imglib2#22]
Would it be possible to Serialize an Img? I don't necessarily mean making Img implement Serializable. 

A more direct question might be: is it possible to get a reference to an Img's type access? For instance, if I have an [ArrayImg<IntType>](https://github.com/imagej/imglib/blob/master/core/src/main/java/net/imglib2/img/array/ArrayImg.java), I'd like to be able to get a reference to the [IntArray](https://github.com/imagej/imglib/blob/master/core/src/main/java/net/imglib2/img/basictypeaccess/array/IntArray.java) that backs it. I could Serialize the access (if it is Serializable), and reconstitute the Img from that.

**Comment by kmader** [comment]
This is probably very naive, but is there a block to having everything in ImgLib2 extend Serializable? For Spark and other distributed tools it is essential that everything is serializable. It is clearly not the most efficient method for serializing many objects, particularly with views, manually rewriting everything from ```FloatType``` to ```ArrayImgFactory``` is quite time consuming.

**Comment by dietzc** [comment]
I agree, we need to do something on this front. 

Does it need to be `Serializable` or do you just need a mechanism to write to a stream? Maybe we could add a SerializationService to ImageJ-Ops, or if possible, use SCIFIO Reader/Writer to do that (assuming you can write to a stream there and its fast enough).

what do you think @ctrueden @hinerm @kmader?

**Comment by ctrueden** [comment]
Yeah, my idea was to address binary marshaling and unmarshaling (of a transient-and-high-performance nature such as needed for distributed computing) with a special SCIFIO format. If the SCIFIO architecture itself is somehow not performant enough to do this well, then we have failed and need to address that.

@kmader I second Christian's question: does it _have_ to be `Serializable` specifically? Because if so, that places a lot of constraints on the design all over the SciJava software stack.

**Comment by kmader** [comment]
@ctrueden and @dietzc , strictly speaking they have to implement ```Serializable``` (or ```Externalizable```) to be directly used. I, however, agree that it would be quite inefficient and limiting to make everything Serializable. 

My solution thus far (the same antipattern as I used with IJ1) is shown (for Scala) in this brief readme (https://gist.github.com/kmader/1d64e64621e63d566f67)

Based on this very limited experience, I would say, the base types should implement ```Serializable``` as right now my code is littered with lambda functions to make empty type objects (```() => new FloatType()```) since the functions are Serializable. For large or data-storing objects (```Img, ArrayImg, etc```), a serializing service would be nice and then we could come up with some nicer pattern for transitioning between serialized and unserialized representations. I am too unfamiliar with the tools of ```ArrayImgFactory``` and the like, but potentially they could be just made ```Serializable``` as well.
